File: 1/2020.11.04/541.Reverse String II/java/solution.java
class Solution {
    public String reverseStr(String s, int k) {
        char[] ch = s.toCharArray();
        int n = s.length();
        int i = 0;
        while (i < n) {
            swap(ch, i, Math.min(i+k-1, n-1));
            i += k*2;
        }
        return new String(ch);
    }
    
    private void swap(char[] ch, int start, int end) {
        while (start < end) {
            char tmp = ch[start];
            ch[start++] = ch[end];
            ch[end--] = tmp;
        }
    }
}

File: 1/2020.11.04/541.Reverse String II/python/solution.py
class Solution:
    def reverseStr(self, s: str, k: int) -> str:
        def swap(l, r):
            while l < r:
                arr[l], arr[r] = arr[r], arr[l]
                l += 1
                r -= 1
        arr = list(s)
        n = len(s)
        i = 0
        while i < n:
            swap(i, min(i+k-1, n-1))
            i += k * 2
        return ''.join(arr)
